### PR TITLE
Nodelet Implementation

### PR DESCRIPTION
**Pull request by naveedhd**
- Don't have a launch file to bundle these two together, still have to test them separately
- Processing works inside callback -> need to implement in thread and queue and possible optimize also

### CODE REVIEW

**Comment by stonier** [comment]
Tested, looks good.

Q) Is there any difference in code between node and nodelet? If not...let's drop the node implementation - nodelet + manager (as in your launcher) can replace it.
- Drop `node.cpp`
- Drop `rplidar.launch`
- Rename `rplidar_nodelet.launch` -> `rplidar.launch`
- Squash the commits into one

I've got `gopher_robot` getting ready for the rplidar nodelet over in [gopher@b19ee2c7](https://bitbucket.org/yujinrobot/gopher/src/b19ee2c7533cf9fd9c359be2070ae04fd3994b08/gopher_robot/launch/components/sensors/rplidar.launch.xml?fileviewer=file-view-default)

**Comment by bit-pirate** [comment]
> - Drop node.cpp

I usually leave the node implementation, if it's just a simple wrapper. Benefits are that nodes are easier for debugging and testing and users not utilising nodeletes don't need to spawn a nodelet manager for just one nodelet.

**Comment by stonier** [comment]
Yeah, I would, but the problem is it isn't just a simple wrapper. At the moment alot of code is duplicated, so it's either move all of that into a library, or drop the node.

How are nodes easier for debugging and testing? As far as I know a nodelet manager and single nodelet don't differ at all except for using `NODELET_XXX` instead of `ROS_XXX`. Subsequently, for me, having the node around is just extra maintenance.

**Comment by stonier** [comment]
Ah, which brings up one more job....
- `ROS_XXX` -> `NODELET_XXX` (see the [wiki page](http://wiki.ros.org/nodelet) for info)

**Comment by bit-pirate** [comment]
> How are nodes easier for debugging and testing?

For example, usually our nodes are single-threaded, whereas with nodelets we often spin up a second thread. Running things in GDB is usually simpler with a node then a nodelet. I remember being unable killing the nodelet within GDB with ctrl+c. I guess it's related to how parameters and konsole in/output is passed to the process by GDB.

> At the moment alot of code is duplicated, so it's either move all of that into a library, or drop the node.

How likely can we bring our changes back upstream? I'd like to add this factor to our decision making. Many ROS users don't use nodelets and might hesitate getting involved. So moving things out into a library would be good, since the node/nodelet part will be minimal then and the library would be used in both cases.

**Comment by stonier** [comment]
Ok, I'd like to clarify the situation to determine what is real and what is myth/urban legend. I don't want to support extra mechanisms because of what _might_ be non-issues.

**Comment by stonier** [comment]
> How likely can we bring our changes back upstream? 

At this stage I'd say for this particular repository any change is probably not likely. Their support is almost absent. I'm almost considering asking them if we can take over the responsibility of maintenance (esp. with respect to debs) but I do not wish to do that till we've confirmed we will use their device long term.

**Comment by naveedhd** [comment]
I have fixed the ROS_xyz to NODELET_xyz and squashed with all previous commits.

Right now, I have implemented a node using nodelet loader named `rplidar_node` (and deleted `node` for brevity). So we have both node and nodelet, but code at one location. Please check if that is correct.

File: CMakeLists.txt [hunk 1]
```diff
@@ -23,13 +28,17 @@ include_directories(
...
 
-install(TARGETS rplidarNode rplidarNodeClient
+add_library(rplidar_nodelet src/rplidar_nodelet.cpp ${RPLIDAR_SDK_SRC})
+target_link_libraries(rplidar_nodelet ${catkin_LIBRARIES})
+
+add_executable(rplidar_node src/rplidar_node.cpp)
+target_link_libraries(rplidar_node rplidar_nodelet ${catkin_LIBRARIES})
+
+
+install(TARGETS rplidarNodeClient
```
Comment (stonier) [inline CMakeLists.txt]: Missing some install rules here...

**Comment by stonier** [comment]
Just added some install rules (always remember to do a `--install` to test things!)

Apart from that, looks good, thanks! :neckbeard:

**Pull request merged**